cookbooks/equilibrium_grain_size/plugins/equilibrium_grain_size.cc

```diff
@@ -0,0 +1,1750 @@
+/*
+  Copyright (C) 2014 - 2018 by the authors of the ASPECT code.
+
+  This file is part of ASPECT.
+
+  ASPECT is free software; you can redistribute it and/or modify
+  it under the terms of the GNU General Public License as published by
+  the Free Software Foundation; either version 2, or (at your option)
+  any later version.
+
+  ASPECT is distributed in the hope that it will be useful,
+  but WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+  GNU General Public License for more details.
+
+  You should have received a copy of the GNU General Public License
+  along with ASPECT; see the file doc/COPYING.  If not see
+  <http://www.gnu.org/licenses/>.
+*/
+
+
+#include "equilibrium_grain_size.h"
+#include <aspect/adiabatic_conditions/interface.h>
+#include <aspect/gravity_model/interface.h>
+#include <aspect/utilities.h>
+#include <aspect/initial_temperature/interface.h>
+#include <aspect/initial_temperature/adiabatic_boundary.h>
+#include <aspect/simulator_signals.h>
+
+#include <deal.II/base/quadrature_lib.h>
+#include <deal.II/fe/fe_values.h>
+#include <deal.II/base/signaling_nan.h>
+
+#include <iostream>
+
+using namespace dealii;
+
+namespace aspect
+{
+  namespace MaterialModel
+  {
+    /**
+     * Additional output fields for the viscosity prior to scaling to be added to
+     * the MaterialModel::MaterialModelOutputs structure and filled in the
+     * MaterialModel::Interface::evaluate() function.
+     */
+    template <int dim>
+    class UnscaledViscosityAdditionalOutputs : public NamedAdditionalMaterialOutputs<dim>
+    {
+      public:
+        UnscaledViscosityAdditionalOutputs(const unsigned int n_points)
+          : NamedAdditionalMaterialOutputs<dim>(std::vector<std::string>(1, "unscaled_viscosity"),
+                                                n_points)
+        {}
+    };
+
+    /**
+     * Additional output fields for the the material type, decribing if we are in the
+     * crust/lithosphere/asthenosphere/lower mantle.
+     */
+    template <int dim>
+    class MaterialTypeAdditionalOutputs : public NamedAdditionalMaterialOutputs<dim>
+    {
+      public:
+        MaterialTypeAdditionalOutputs(const unsigned int n_points)
+          : NamedAdditionalMaterialOutputs<dim>(std::vector<std::string>(1, "material_type"),
+                                                n_points)
+        {}
+    };
```

Comment:
These two classes, along with the `FunctorDepthAverageUnscaledViscosity` one below, are purely local to this file. It might be nice to surround their definition with `namespace { ... };`, i.e., put them into an "anonymous namespace".